src/coffee/stockimport.coffee

```diff
@@ -179,19 +177,94 @@ class StockImport
           stocks.push d
     stocks
 
-  _mapStockFromCSV: (rows, skuIndex = 0, quantityIndex = 1) ->
-    _.map rows, (row) =>
-      sku = row[skuIndex].trim()
-      quantity = row[quantityIndex]?.trim()
-      @_createInventoryEntry sku, quantity
+  _mapStockFromCSV: (rows, mappedHeaderIndexes) ->
+    return new Promise (resolve, reject) =>
+      rowIndex = 0 # very weird that csv does not support this internally
+      csv.transform(rows,(row, cb) =>
+        rowIndex++
+        _data = {}
+
+        Promise.each(row, (cell, index) =>
+          headerName = mappedHeaderIndexes[index]
+
+          if CONS.HEADER_CUSTOM_REGEX.test headerName
+            customTypeKey = row[mappedHeaderIndexes.indexOf(CONS.HEADER_CUSTOM_TYPE)]
+
+            @_getCustomTypeDefinition(customTypeKey).then (response) =>
+              customTypeDefinition = response.body
+              @_mapCustomField(_data, cell, headerName, customTypeDefinition, rowIndex)
+
+          else
+            Promise.resolve(@_mapCellData(cell, headerName)).then (cellData) =>
+              _data[headerName] = cellData
+
+        ).then =>
+          if _.size(@customFieldMappings.errors) isnt 0
+            return cb @customFieldMappings.errors
+          cb null, _data
+      , (err, data) ->
+        if err
+          reject(err)
+        resolve(data)
+      )
+
+
+  _mapCellData: (data, headerName) ->
+    data = data?.trim()
+    switch on
+      when CONS.HEADER_QUANTITY is headerName then parseInt(data, 10) or 0
+      when CONS.HEADER_RESTOCKABLE is headerName then parseInt(data, 10)
+      when CONS.HEADER_SUPPLY_CHANNEL is headerName then @_mapChannelKeyToReference data
+      else data
+
+  _mapCustomField: (data, cell, headerName, customTypeDefinition, rowIndex) ->
+    fieldName = headerName.split(CONS.HEADER_CUSTOM_SEPERATOR)[1]
+    lang = headerName.split(CONS.HEADER_CUSTOM_SEPERATOR)[2]
+
+    # set data.custom once per row with the type defined
+    if !data.custom
+      data.custom = {
+        "type": {
+          "id": customTypeDefinition.id
+        },
+        "fields": {}
+      }
+    # Set localized object if present
+    if lang
+      data.custom.fields[fieldName] =
+        _.defaults (data.custom.fields[fieldName] || {}),
+        @customFieldMappings.mapFieldTypes customTypeDefinition.fieldDefinitions,customTypeDefinition.key,rowIndex,fieldName,cell,lang
+    else
+      data.custom.fields[fieldName] = @customFieldMappings.mapFieldTypes customTypeDefinition.fieldDefinitions,customTypeDefinition.key,rowIndex,fieldName,cell
+
+
+  _getCustomTypeDefinition: _.memoize (customTypeKey) ->
+    @__getCustomTypeDefinition customTypeKey
```

Comment:
Can you explain how `memoize` help in this case? Not just for me but in the code also, for future references ;)